rfcs/0001-messages.md

```diff
@@ -0,0 +1,397 @@
+}
+
+/// The data required to serialize `frost::SigningPackage`.
+///
+/// The aggregator decides what message is going to be signed and
+/// sends it to each signer with all the commitments collected.
+struct messages::SigningPackage {
+    /// The message to be signed: `frost::SigningPackage.message`
+    message: Vec<u8>,
+    /// The collected commitments for each signer as a hashmap of
+    /// unique participant identifiers: `frost::SigningPackage.signing_commitments`
+    ///
+    /// Signing packages that contain duplicate or missing `ParticipantId`s are invalid.
+    signing_commitments: HashMap<ParticipantId, SigningCommitments>,
+}
+
+/// The data required to serialize `frost::SignatureShare`.
+///
+/// Each signer sends their signatures to the aggregator who is going to collect them
+/// and generate a final spend signature.
+struct messages::SignatureShare {
+     /// This participant's signature over the message: `frost::SignatureShare.signature`
+    signature: frost::Scalar,
+}
+
+/// The data required to serialize a successful output from `frost::aggregate()`.
+///
+/// The final signature is broadcasted by the aggregator to all signers.
+struct messages::AggregateSignature {
+    /// The aggregated group commitment: `Signature<SpendAuth>.r_bytes` returned by `frost::aggregate`
+    group_commitment: jubjub::AffinePoint,
+    /// A plain Schnorr signature created by summing all the signature shares:
+    /// `Signature<SpendAuth>.s_bytes` returned by `frost::aggregate`
+    schnorr_signature: frost::Scalar,
+}
+```
+
+## Validation
+
+Validation is implemented to each new data type as needed. This will ensure the creation of valid messages before they are send and right after they are received. We create a trait for this as follows:
+
+```rust
+pub trait Validate {
+    fn validate(&self) -> Result<&Self, MsgErr>;
+}
+```
+
+And we implement where needed. For example, in the header, sender and receiver can't be the same:
+
+```rust
+impl Validate for Header {
+    fn validate(&self) -> Result<&Self, MsgErr> {
+        if self.sender.0 == self.receiver.0 {
+            return Err(MsgErr::SameSenderAndReceiver);
+        }
+        Ok(self)
+    }
+}
+```
+
+This will require to have validation error messages as:
+
+```rust
+use thiserror::Error;
+
+#[derive(Clone, Error, Debug)]
+pub enum MsgErr {
+    #[error("sender and receiver are the same")]
+    SameSenderAndReceiver,
+}
+```
+
+Then to create a valid `Header` in the sender side we call:
+
+```rust
+let header = Validate::validate(&Header {
+    ..
+}).expect("a valid header");
+```
+
+The receiver side will validate the header using the same method. Instead of panicking the error can be ignored to don't crash and keep waiting for other (potentially valid) messages.
+
+```rust
+if let Ok(header) = msg.header.validate() {
+    ..
+}
+```
+
+### Rules
+
+The following rules must be implemented:
+
+#### Header
+
+- `msg_type` must be a known `MsgType` value.
```

Comment:
I think we can remove this rule as we don't have this field any more in the rust side. We can add this as a note in serialization section, as it needs to be implemented in the deserialization code.

Reply:
Sure, go for it!